src/froll.c

```diff
@@ -406,11 +405,253 @@ void frollsumExact(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool
   }
 }
 
+inline void windowmax(double *x, uint64_t o, int k, double *w, uint64_t *iw) {
+  for (int i=0; i<k-1; i++) {
+    //Rprintf("windowmax iteration %d, offset %d, first x val %f, testing x[o+i-k+1] >= w[0]: x[%d-%d+1] >= w[0]: %f >= %f: %d\n", i, o, x[o], i, k, x[o+i-k+1], w[0], x[o+i-k+1] >= w[0]);
+    if (x[o+i-k+1] >= w[0]) { // what if that is never satisfied? test!
+      iw[0] = o+i-k+1;
+      w[0] = x[iw[0]];
+    }
+  }
+}
+inline void windowmaxnarm(double *x, uint64_t o, int k, bool narm, int *nc, double *w, uint64_t *iw) {
+  for (int i=0; i<k-1; i++) {
+    //Rprintf("windowmax iteration %d, offset %d, first x val %f, testing x[o+i-k+1] >= w[0]: x[%d-%d+1] >= w[0]: %f >= %f: %d\n", i, o, x[o], i, k, x[o+i-k+1], w[0], x[o+i-k+1] >= w[0]);
+    if (R_FINITE(x[o+i-k+1])) {
+      if (x[o+i-k+1] >= w[0]) { // what if that is never satisfied? test!
+        iw[0] = o+i-k+1;
+        w[0] = x[iw[0]];
+      }
+    } else if (narm) {
+      nc[0]++;
+    } else {
+      w[0] = NA_REAL;
+    }
+  }
+}
+/* fast rolling max - fast
+ * fast online algorithm do single pass over elements keeping track of recent max and its index
+ * if index of max is within progressing window then it keeps running single pass
+ * whenever max is leaving the window (index of max is outside of iterator minus window size) then new maximum is computed via nested loop on current complete window
+ * new max is used to continue outer single pass as long as new max index is not leaving the running window
+ * should scale well for bigger window size, may carry overhead for small window, needs benchmarking
+ * TODO NA handling
+ */
+void frollmaxFast(double *x, uint64_t nx, ans_t *ans, int k, double fill, bool narm, int hasna, bool verbose) {
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: running for input length %"PRIu64", window %d, hasna %d, narm %d\n"), "frollmaxFast", (uint64_t)nx, k, hasna, (int)narm);
+  double w = R_NegInf; // window max
+  uint64_t cmax = 0; // counter of nested loops for verbose
+  uint64_t iw = 0; // index of window max
+  bool truehasna = hasna>0;
+  if (!truehasna) {
+    int i;
+    for (i=0; i<k-1; i++) { // #loop_counter_not_local_scope_ok
+      if (!R_FINITE(x[i])) {
+        truehasna = true;
+        break;
+      }
+      if (x[i] >= w) { // >= rather than > because we track most recent maximum using iw
+        iw = i;
+        w = x[iw];
+      }
+      ans->dbl_v[i] = fill;
+    }
+    if (!truehasna && !R_FINITE(x[i])) {
+      if (x[i] >= w) {
+        iw = i;
+        w = x[iw];
+      } else {
+        // ensure iw ok, case when all window was -Inf or NA?
+      }
+      ans->dbl_v[i] = w;
+    } else {
+      truehasna = true;
+    }
```

Comment:
again, can't we move this into the loop?

Reply:
Loop goes over `1:length(x)` (excluding trailing incomplete window which was taken out also for speed) therefore we want to take out from the loop as much as possible. Making tight loops as light as possible was a priority in rolling functions and as you can see there are multiple places like that. I used to branch outside of the loop and create each loop for each branch. Every single branch inside the loop counts. These are microseconds but we don't want to be lagging behind other tools, see https://github.com/jangorecki/rollbench